### PR TITLE
Extract chorus visualization into VisLegacy component (REVIEW 1st, MERGE LAST)

### PR DESCRIPTION
**Pull request by kevin-alpine**
Curran, Sara, & Lawrence asked me to help Curran modify Chorus's visualization system to work with Curran's Chiasm project.

The goal of this work is to reach parity: we're simply swapping out the current system to use Chiasm for visualization.

As is often the case -- even though, during the meeting we had on Tuesday, this was the least ambitious path (adding no new functionality, just using Chiasm to replace the current Dataset visualization system), as I looked into accomplishing it on Tuesday night, I was stunned by how big of a system it is really is: there are **~22k lines of code** in Chorus devoted to generating those graphs.

So -- in order to do it in a sane way, I'm doing in four steps:
* 1st (this step): extract all the code currently in the system into a component called VisLegacy.
* 2nd: duplicate VisLegacy into a 2nd component called VisChiasm and switch over Chorus to use that new component.
* 3rd: aggressively modify the VisChiasm component to work with Chiasm and clean out all the bits of VisLegacy that are now deprecated.
* 4th (later): Delete the VisLegacy component.

In order to make these changes reviewable, I will create a chain of Github PRs (each merges into one another before finally merging into master).

@midnighteuler, @AndrewLngdn, @jleealpine, @curran: this branch is cool because it's a validation of how useful the "components" idea can be -- also, I needed to be able to divide up the Backbone code into components, so I had to invent a simple way to do it:
* I divided the `Messages_en.properties` file into two pieces.  I store the URLs in an array, and load and concatenate them before passing them into the I18n js.
* This plugin is not namespaced -- [it's "Full" but not "Mountable"](http://guides.rubyonrails.org/engines.html) ... so, once the app boots ... everything gets mashed together into one namespace.  This doesn't get us quite as much namespacing and code isolation but it does get us swap-ability.

I will comment more in the body of the PR ...

### CODE REVIEW

**Comment by alpinebuilder** [comment]
Can one of the admins verify this patch?

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -43,11 +43,18 @@
 //= require csrf
 //= require application_misc
 //= require chorus
+
+// KT: Some 'componentized' javascript needs to be loaded before other files (mixins & translations, for example) ...
+// put these in '<component name>_boot' and require it up top in the manifest file:
+//= require vis_legacy_boot
```
Comment (kevin-alpine) [inline app/assets/javascripts/application.js]: Some 'componentized' javascript needs to be loaded before other files (mixins & translations, for example) ... put these in `<component name>_boot` and require it up top in the manifest file ...

File: app/assets/javascripts/application.js [hunk 2]
```diff
@@ -102,6 +109,7 @@
 //= require pages/dataset/workspace_dataset_show_page
 //= require pages/search_index_page
 //= require pages/workspaces/workspace_show_page
+
 //= require_tree .
 
-//= require vis_engine/application
+//= require vis_legacy
```
Comment (kevin-alpine) [inline app/assets/javascripts/application.js]: The rest of the componentized JS can be required down here via the component's JS manifest file ...

File: app/assets/javascripts/chorus.js [hunk 3]
```diff
@@ -2,7 +2,9 @@ window.Chorus = function chorus$Global() {
     var self = this;
     self.models = {};
     self.views = {};
-    self.views.visualizations = {};
+
+    self.translation_files = ['/assets/messages_en.properties'];
```
Comment (kevin-alpine) [inline app/assets/javascripts/chorus.js]: The "main" `messages_en.properties`

File: components/vis_legacy/app/assets/javascripts/vis_legacy_boot.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+//= require_tree ./mixins
+//= require_self
+
+window.chorus.translation_files.push('/assets/visualizations_messages_en.properties');
```
Comment (kevin-alpine) [inline components/vis_legacy/app/assets/javascripts/vis_legacy_boot.js]: `visualizations_messages_en.properties`

File: app/assets/javascripts/utilities/translation.js [hunk 5]
```diff
@@ -22,18 +22,22 @@ chorus.translation = {
...
+})).done(function() {
+
+    var concatenatedTranslationsFiles = '';
+    for (var i = 0; i < arguments.length; i++) {
+        concatenatedTranslationsFiles += arguments[i][0];
+    }
+
     I18n.translations = {};
-    I18n.translations[chorus.locale] = chorus.translation.parseProperties(data);
+    I18n.translations[chorus.locale] = chorus.translation.parseProperties(concatenatedTranslationsFiles);
```
Comment (kevin-alpine) [inline app/assets/javascripts/utilities/translation.js]: Load and concatenate all the `messenges.properties` files ...

File: app/assets/javascripts/views/dataset/dataset_content_details.js [hunk 6]
```diff
@@ -1,8 +1,9 @@
 chorus.views.DatasetContentDetails = chorus.views.Base.include(
-        chorus.Mixins.StickyHeader
+        chorus.Mixins.StickyHeader,
+        chorus.Mixins.DatasetContentDetailsVisualizations
```
Comment (kevin-alpine) [inline app/assets/javascripts/views/dataset/dataset_content_details.js]: We can pull code out of this (nonshared) Backbone file ... and move it into the component by using [Backbone mixins](http://ricostacruz.com/backbone-patterns/#mixins) ...

File: app/assets/javascripts/views/dataset/dataset_content_details.js [hunk 7]
```diff
@@ -231,171 +173,40 @@ chorus.views.DatasetContentDetails = chorus.views.Base.include(
...
-              "state": {
-                "bindings": [
-                  "dataLoader.data -> dataReduction.dataIn",
-                  "dataReduction.dataOut -> visualization.data"
-                ]
-              }
-            }
-        };
-
-        chiasm.setConfig(config);
```
Comment (kevin-alpine) [inline app/assets/javascripts/views/dataset/dataset_content_details.js]: I will put back Curran's code in step 3 -- for now, I'm just restoring the 5.6.1 behavior.

File: app/assets/javascripts/views/dataset/dataset_content_details.js [hunk 8]
```diff
@@ -231,171 +173,40 @@ chorus.views.DatasetContentDetails = chorus.views.Base.include(
...
-        this.$(".chart_type_title." + type).removeClass("hidden");
+        this.$('.chart_type_title').addClass('hidden');
+        var type = this.$('.selected').data('chart_type');
+        this.$('.chart_type_title.' + type).removeClass('hidden');
     },
 
     additionalContext: function() {
         var workspaceArchived = this.options.workspace && !this.options.workspace.isActive();
-        return {
+        return _.extend(this.additionalContextForVisualizations(), {
```
Comment (kevin-alpine) [inline app/assets/javascripts/views/dataset/dataset_content_details.js]: You can extract the import 'additionalContext' methods into a component specific mixin using this method ... I copied this from [tags_context.js](https://github.com/Chorus/chorus/blob/master/app/assets/javascripts/mixins/tags_context.js#L3) ...

File: app/assets/stylesheets/application.scss [hunk 9]
```diff
@@ -24,10 +24,12 @@
...
  *= require users
- *= require visualizations
  *= require widgets
  *= require workfiles
  *= require worklets
  *= require workspaces
  *= require help
+
+ *= require vis_legacy
+
```
Comment (kevin-alpine) [inline app/assets/stylesheets/application.scss]: Load the component specific styles via component scss manifest file ...

File: app/assets/stylesheets/dataset/_dataset_content_details.scss [hunk 10]
```diff
@@ -1,11 +1,4 @@
-@mixin action_bar_icon($name) {
-    .#{$name}.action_bar_icon {
-        background-image: url("/images/visualizations/create_icons/#{$name}_off.png");
-        &.selected {
-            background-image: url("/images/visualizations/create_icons/#{$name}_on.png");
-        }
-    }
-}
```
Comment (kevin-alpine) [inline app/assets/stylesheets/dataset/_dataset_content_details.scss]: A "sass" mixin ... references a URL that has been moved inside of the VisLegacy component ... so, we extract it ...

File: app/controllers/attachments_controller.rb [hunk 11]
```diff
@@ -2,13 +2,17 @@ class AttachmentsController < ApplicationController
...
       attachment_content = params[:contents]
     else
-      transcoder = SvgToPng.new(params[:svg_data])
-      attachment_content = transcoder.fake_uploaded_file(params[:file_name])
+      if defined?(VisLegacy::Engine)
+        transcoder = SvgToPng.new(params[:svg_data])
+        attachment_content = transcoder.fake_uploaded_file(params[:file_name])
+      else
+        raise "No Visualization component attached."
+      end
```
Comment (kevin-alpine) [inline app/controllers/attachments_controller.rb]: This is a reverse reference, so we raise an exception if there's no VisLegacy component defined ... there's not really any "default" behavior we could implement here, so make it noisy if it's broken.

File: components/vis_legacy/app/decorators/models/chorus_workfile_decorator.rb [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+  def svg_data=(svg_data)
+    # Note: This conversion upon assignment depends on the "file_name"
+    # being populated--and that file_name is being assigned from the same
+    # hash (during build_for)
+    # So 'file_name' has to precede 'svg_data' in the params hash
+    # providing this.
+    transcoder = SvgToPng.new(svg_data)
+    versions.build :contents => transcoder.fake_uploaded_file(file_name)
+  end
+end
```
Comment (kevin-alpine) [inline components/vis_legacy/app/decorators/models/chorus_workfile_decorator.rb]: Note how we're hooking in to `ChorusWorkfile` and adding this method via `.class_eval` ...

**Comment by curran** [comment]
Just to keep a record in case we want to use this database random sampling API in the future - here's the front end Chiasm module that connects to it:

```javascript
//A Chiasm plugin that loads data from the vis_chiasm random sampling API.
var Model = require("model-js");
var ChiasmComponent = require("chiasm-component");
var d3 = require("d3");
var dsvDataset = require("dsv-dataset");

module.exports = function (){

  var my = ChiasmComponent({
    dataset_id: Model.None,
    numRows: 1000
  });

  my.when(["dataset_id", "numRows"], function (dataset_id, numRows){
    if(dataset_id !== Model.None){

      var columnsPath = "datasets/" + dataset_id + "/chiasm_api_datasets/show_column_data";
      var dataPath = "datasets/" + dataset_id + "/chiasm_api_datasets/show_data";

      dataPath += "?numRows=" + numRows;

      d3.json(columnsPath, function(error, columns) {
        if(error){ throw error; }
        my.metadata = { columns: columns };
      });

      d3.xhr(dataPath, function (error, xhr){
        if(error){ throw error; }
        my.dsvString = xhr.response;
      });
    }
  });

  my.when(["dsvString", "metadata"], function (dsvString, metadata){
    my.dataset = dsvDataset.parse({
      dsvString: dsvString, 
      metadata: metadata
    });
    my.data = my.dataset.data;
  });

  return my;
}
```

**Pull request merged**